src/app/components/chat/window/output/AppChatWindowOutputItem.vue

```diff
@@ -0,0 +1,530 @@
+	Popper.hideAll();
+}
+
+async function removeMessage() {
+	Popper.hideAll();
+
+	const result = await ModalConfirm.show(
+		$gettext(`Are you sure you want to remove this message?`)
+	);
+
+	if (!result) {
+		return;
+	}
+
+	setMessageEditing(chat.value, null);
+	chatRemoveMessage(chat.value, room.value, message.value.id);
+}
+
+let canClearFocus = false;
+let isFocused = false;
+
+async function onFocusItem() {
+	isFocused = true;
+	// Allow a second click to blur only after ~1/2 of our animations play.
+	setTimeout(() => (canClearFocus = isFocused), 100);
+}
+
+function onBlurItem() {
+	isFocused = false;
+	canClearFocus = false;
+}
+
+async function onClickItem() {
+	if (isFocused && canClearFocus) {
+		itemWrapper.value?.blur();
+	}
+}
+
+function onItemWrapperResize() {
+	if (!itemWrapper.value) {
+		return;
+	}
+
+	const { offsetWidth } = itemWrapper.value;
+	const decoratorPadding = 12 * 2;
+	contentViewerBounds.width = Math.max(offsetWidth - decoratorPadding, 100);
+}
+</script>
+
+<template>
+	<div
+		ref="root"
+		class="chat-window-output-item"
+		:class="{
+			'-message-queued': message._showAsQueued,
+			'-message-editing': isEditing,
+		}"
+	>
+		<a v-if="message.showAvatar" class="-avatar">
+			<AppPopper placement="right">
+				<img
+					class="-avatar-img img-responsive"
+					:src="message.user.img_avatar"
+					alt=""
+					draggable="false"
+				/>
+
+				<template #popover>
+					<AppChatUserPopover :user="message.user" :room="room" />
+				</template>
+			</AppPopper>
+		</a>
+
+		<div
+			ref="itemWrapper"
+			v-app-observe-dimensions="onItemWrapperResize"
+			class="-item-container-wrapper"
+			tabindex="-1"
+			@focus="onFocusItem"
+			@blur="onBlurItem"
+			@click.capture="onClickItem"
+		>
+			<div class="-item-container">
+				<div
+					class="-item-decorator"
+					:style="{
+						'background-image': isEditingColor
+							? `linear-gradient(${isEditingColor}, ${isEditingColor}`
+							: undefined,
+					}"
+				>
+					<div v-if="message.showMeta" class="-item-byline">
+						<RouterLink class="-user link-unstyled" :to="message.user.url">
+							{{ message.user.display_name }}
+						</RouterLink>
+						<span class="-username"> @{{ message.user.username }} </span>
+						<span>
+							<AppChatWindowOutputItemDetails
+								:message="message"
+								:room="room"
+								:timestamp-margin-left="8"
+								show-am-pm
+							/>
+						</span>
+					</div>
+
+					<AppContentViewer
+						:source="message.content"
+						:display-rules="displayRules"
+						:parent-bounds="contentViewerBounds"
+					/>
+
+					<span
+						v-if="editingState"
+						v-app-tooltip.touchable="editingState.tooltip"
+						class="-edited"
+						:class="{ 'text-muted': !isEditing }"
+					>
+						<AppTranslate>{{ editingState.display }}</AppTranslate>
+					</span>
+				</div>
+
+				<div v-if="!message.showMeta" class="-floating-data-left">
+					<AppChatWindowOutputItemDetails :message="message" :room="room" />
```

Comment:
on mobile when there's an error, this will be hidden behind a "tap on message" before being able to retry?

Reply:
Changed to behave more like app, where failed and sending messages display that state within the message itself (pinned to bottom) instead of as some overlay content.